zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala

```diff
@@ -611,48 +590,66 @@ private[consumer] final class Runloop(
       _ => command.succeed.as(state.copy(subscription = command.subscription))
     )
 
-  def run: ZIO[Scope, Nothing, Fiber.Runtime[Throwable, Unit]] =
-    ZStream
-      .mergeAll(3, 1)(
-        ZStream.repeatWithSchedule(Command.Poll, Schedule.fixed(pollFrequency)),
-        ZStream.fromQueue(requestQueue).mapChunks(c => Chunk.single(Command.Requests(c))),
-        ZStream.fromQueue(commandQueue)
-      )
-      .tap(cmd => diagnostics.emitIfEnabled(DiagnosticEvent.RunloopEvent(cmd)))
-      .runFoldZIO(State.initial) { (state, cmd) =>
-        ZIO.ifZIO(isShutdown)(onTrue = handleShutdown(state, cmd), onFalse = handleOperational(state, cmd))
-      }
+  def run: ZIO[Scope, Nothing, Fiber.Runtime[Throwable, Any]] = {
+    def processCommands(state: State, wait: Boolean): Task[State] = for {
+      commands <- if (wait)
+                    commandQueue.takeBetween(1, commandQueueSize).timeoutTo(Chunk.empty)(identity)(pollFrequency)
+                  else commandQueue.takeAll // Gather available commands or return immediately if nothing in the queue
+
+      isShutdown <- isShutdown
+      updatedState <- ZIO.foldLeft(commands)(state) { case (s, cmd) =>
+                        (if (isShutdown) handleShutdown(s, cmd) else handleOperational(s, cmd)) <*
+                          diagnostics.emitIfEnabled(DiagnosticEvent.RunloopEvent(cmd))
```

Comment:
What do you guys think about this change:
```suggestion
        isShutdown <- isShutdown
        commands <- if (!wait || isShutdown)
                      commandQueue.takeAll // Gather available commands or return immediately if nothing in the queue
                    else commandQueue.takeBetween(1, commandQueueSize).timeoutTo(Chunk.empty)(identity)(pollFrequency)

        updatedState <- ZIO.foldLeft(commands)(state) { case (s, cmd) =>
                          (if (isShutdown) handleShutdown(s, cmd) else handleOperational(s, cmd)) <*
                            diagnostics.emitIfEnabled(DiagnosticEvent.RunloopEvent(cmd))
                        }
```

So that if we're shutting down, we take all the commands and end them.

Reply:
This shutdown is really a 'gracefully stop fetching new data', so there may still be commits being made. We need to process those commit commands, which may come in over several iterations of the poll loop.